client/client-multi/client-c/src/platform-impl/kaa_tcp_channel.c

```diff
@@ -0,0 +1,1349 @@
+kaa_error_t kaa_tcp_channel_set_access_point(void *context, kaa_access_point_t *access_point);
+
+/*
+ * Parser handlers
+ */
+void kaa_tcp_channel_connack_message_callback(void *context, kaatcp_connack_t message);
+void kaa_tcp_channel_disconnect_message_callback(void *context, kaatcp_disconnect_t message);
+void kaa_tcp_channel_kaasync_message_callback(void *context, kaatcp_kaasync_t *message);
+void kaa_tcp_channel_pingresp_message_callback(void *context);
+
+/*
+ * Internal functions
+ */
+kaa_error_t kaa_tcp_channel_socket_io_error(kaa_tcp_channel_t *self);
+kaa_error_t kaa_tcp_channel_authorize(kaa_tcp_channel_t *self);
+bool is_service_pending(kaa_tcp_channel_t *self, const kaa_service_t service);
+kaa_error_t kaa_tcp_channel_delete_pending_services(kaa_tcp_channel_t *self, const kaa_service_t services[], size_t service_count);
+kaa_error_t kaa_tcp_channel_update_pending_services(kaa_tcp_channel_t *self, const kaa_service_t services[], size_t service_count);
+kaa_error_t kaa_tcp_channel_set_access_point_hostname_resolved(void *context, const kaa_sockaddr_t *addr, kaa_socklen_t addr_size);
+kaa_error_t kaa_tcp_channel_set_access_point_hostname_resolve_failed(void *context);
+uint32_t get_uint32_t(const char *buffer);
+kaa_error_t kaa_tcp_channel_connect_access_point(kaa_tcp_channel_t *self);
+kaa_error_t kaa_tcp_channel_release_access_point(kaa_tcp_channel_t *self);
+kaa_error_t kaa_tcp_channel_write_pending_services(kaa_tcp_channel_t *self, kaa_service_t *service, size_t services_count);
+kaa_error_t kaa_tcp_write_buffer(kaa_tcp_channel_t *self);
+char* kaa_tcp_write_pending_services_allocator_fn(void *context, size_t buffer_size);
+kaa_error_t kaa_tcp_channel_ping(kaa_tcp_channel_t *self);
+kaa_error_t kaa_tcp_channel_disconnect_internal(kaa_tcp_channel_t *self, kaatcp_disconnect_reason_t return_code);
+
+/*
+ * Create TCP channel object
+ */
+kaa_error_t kaa_tcp_channel_create(kaa_transport_channel_interface_t *self, kaa_logger_t *logger, kaa_service_t *supported_services, size_t supported_service_count)
+{
+    KAA_RETURN_IF_NIL2(self, logger, KAA_ERR_BADPARAM);
+
+    KAA_LOG_TRACE(logger, KAA_ERR_NONE, "Kaa TCP channel creating....");
+
+    kaa_error_t error_code = KAA_ERR_NONE;
+
+    kaa_tcp_channel_t *kaa_tcp_channel = (kaa_tcp_channel_t *) KAA_CALLOC(1, sizeof(kaa_tcp_channel_t));
+    KAA_RETURN_IF_NIL(kaa_tcp_channel, KAA_ERR_NOMEM);
+
+    kaa_tcp_channel->channel_state = KAA_TCP_CHANNEL_UNDEFINED;
+    kaa_tcp_channel->access_point.state = AP_NOT_SET;
+    kaa_tcp_channel->access_point.socket_descriptor = KAA_TCP_SOCKET_NOT_SET;
+
+    if (supported_service_count > 0) {
+        kaa_tcp_channel->supported_services = (kaa_service_t *) KAA_MALLOC(supported_service_count * sizeof(kaa_service_t));
+        if (!kaa_tcp_channel->supported_services) {
+            kaa_tcp_channel_release_context(kaa_tcp_channel);
+            return KAA_ERR_NOMEM;
+        }
+
+        int i = 0;
```

Comment:
Please, change type to size_t to avoid compilation warnings. Also you may declare the i counter in the loop body.